### PR TITLE
Measure model integration

### PR DESCRIPTION
**Pull request by mayerm94**
Transition to using the new measure model from cqm-models
Pull requests into cqm-execution require the following. Submitter and reviewer should :white_check_mark: when done. For items that are not-applicable, note it's not-applicable ("N/A") and :white_check_mark:.

**Submitter:**
- [x] This pull request describes why these changes were made.
- [x] Internal ticket for this PR:  https://jira.mitre.org/browse/BONNIE-1607
- [x] Internal ticket links to this PR
- [x] Code diff has been done and been reviewed
- [x] Tests are included and test edge cases
- [x] Tests have been run locally and pass

**Reviewer 1:**

Name:
- [x] Code is maintainable and reusable, reuses existing code and infrastructure where appropriate, and accomplishes the task’s purpose
- [x] The tests appropriately test the new code, including edge cases
- [x] You have tried to break the code

**Reviewer 2:**

Name: @losborne 
- [x] Code is maintainable and reusable, reuses existing code and infrastructure where appropriate, and accomplishes the task’s purpose
- [x] The tests appropriately test the new code, including edge cases
- [x] You have tried to break the code

### CODE REVIEW

File: lib/helpers/calculator_helpers.js [hunk 1]
```diff
@@ -18,41 +18,36 @@ module.exports = class CalculatorHelpers {
...
       );
 
       // initialize population counts
-      Measure.ALL_POPULATION_CODES.forEach((popCode) => {
-        if (population[popCode]) {
-          if (popCode === 'OBSERV') {
-            populationResults.values = [];
-          } else {
+      Object.keys(populationSet.populations).forEach((popCode) => {
+        if (popCode[0] != '_'){
```
Comment (jbradl11) [inline lib/helpers/calculator_helpers.js]: Add a comment about why we need to check for "_" (What does it mean)

File: lib/helpers/calculator_helpers.js [hunk 2]
```diff
@@ -168,61 +163,57 @@ module.exports = class CalculatorHelpers {
...
-    const cqlMap = measure.populations_cql_map;
+
     // Loop over all population codes ("IPP", "DENOM", etc.)
-    Measure.ALL_POPULATION_CODES.forEach((popCode) => {
+    Object.keys(populationSet.populations).forEach((popCode) => {
       let value;
-      if (cqlMap[popCode]) {
-        // Array of populations defined for the given popCode
-        const definedPops = cqlMap[popCode];
+      if (popCode[0] == '_') return;
```
Comment (jbradl11) [inline lib/helpers/calculator_helpers.js]: Similar to above comment.

File: lib/helpers/calculator_helpers.js [hunk 3]
```diff
@@ -237,62 +228,51 @@ module.exports = class CalculatorHelpers {
    * @returns {Object} The episode results. Map of episode id to population results which is a map of "POPNAME"
    * to Integer result. Except for OBSERVs, their key is 'value' and value is an array of results.
    */
-  static createEpisodePopulationValues(measure, population, results, patientId, observationDefs, populationIndex) {
+  static createEpisodePopulationValues(measure, populationSet, patientResults, observationDefs) {
```
Comment (jbradl11) [inline lib/helpers/calculator_helpers.js]: Update comments above this function with the parameter changes

File: lib/helpers/calculator_helpers.js [hunk 4]
```diff
@@ -18,41 +18,36 @@ module.exports = class CalculatorHelpers {
    *   their key is 'value' and value is an array of results. Second result is the the episode results keyed by
    *   episode id and with the value being a set just like the patient results.
    */
-  static createPopulationValues(measure, population, results, patientId, observationDefs, populationIndex) {
+  static createPopulationValues(measure, populationSet, patientResults, observationDefs) {
```
Comment (losborne) [inline lib/helpers/calculator_helpers.js]: With `population` being changed to `populationSet`, we have everything we need to build `observationDefs` within this function rather than passing it in as a parameter.  Currently, with all observations being the same, this isn't needed, but `observationDefs` is locked to `measure.population_sets[0]` in calculator.js, so if they are different in the future, it will need to be moved into this function instead.

File: lib/helpers/calculator_helpers.js [hunk 5]
```diff
@@ -168,61 +163,57 @@ module.exports = class CalculatorHelpers {
    * @returns {Object} The population results. Map of "POPNAME" to Integer result. Except for OBSERVs,
    *   their key is 'value' and value is an array of results.
    */
-  static createPatientPopulationValues(measure, population, results, patientId, observationDefs, populationIndex) {
+  static createPatientPopulationValues(measure, populationSet, patientResults, observationDefs) {
```
Comment (losborne) [inline lib/helpers/calculator_helpers.js]: Similar comments here as above regarding observationDefs

File: lib/helpers/calculator_helpers.js (hunk 4 above)
Comment (mayerm94) [reply lib/helpers/calculator_helpers.js]: If this observationDef generation was added to this function, and we generated the observationsDefs for each population_set, then we would be adding duplicate copies of the generated elm onto the mainLibraryElm. The observations in measure.population_sets[0] are the same in every single population_set

`mainLibraryElm.library.statements.def.push(generatedELMJSON);`

File: lib/helpers/calculator_helpers.js [hunk 6]
```diff
@@ -9,50 +8,42 @@ module.exports = class CalculatorHelpers {
...
-      Measure.ALL_POPULATION_CODES.forEach((popCode) => {
-        if (population[popCode]) {
-          if (popCode === 'OBSERV') {
-            populationResults.values = [];
-          } else {
-            populationResults[popCode] = 0;
-          }
+      Object.keys(populationSet.populations).forEach((popCode) => {
+        // Mongoose adds keys that start with '_' we do not care about these
+        if (popCode[0] !== '_') {
```
Comment (jbradl11) [inline lib/helpers/calculator_helpers.js]: I think we should spend some time on a different way to handle this. Ideally, they wouldn't be included in these structures.

File: lib/helpers/calculator_helpers.js [hunk 7]
```diff
@@ -65,7 +56,8 @@ module.exports = class CalculatorHelpers {
             popResult.forEach((value) => {
               populationResults.values.push(value);
             });
-          } else {
+          // Mongoose adds keys that start with '_' we do not care about these
+          } else if (popCode[0] !== '_') {
```
Comment (jbradl11) [inline lib/helpers/calculator_helpers.js]: See comment above.

File: lib/helpers/calculator_helpers.js [hunk 8]
```diff
@@ -159,183 +151,162 @@ module.exports = class CalculatorHelpers {
...
-      let value;
-      if (cqlMap[popCode]) {
-        // Array of populations defined for the given popCode
-        const definedPops = cqlMap[popCode];
 
-        const cqlPopulation = this.getPopulationFromIndex(definedPops, populationIndex, population, popCode);
+    // Loop over all population codes ("IPP", "DENOM", etc.)
+    Object.keys(populationSet.populations).forEach((popCode) => {
+      // Mongoose adds keys that start with '_' we do not care about these
+      if (popCode[0] === '_') return;
```
Comment (jbradl11) [inline lib/helpers/calculator_helpers.js]: See comment above

File: lib/helpers/calculator_helpers.js [hunk 9]
```diff
@@ -159,183 +151,162 @@ module.exports = class CalculatorHelpers {
...
 
-    popCodesInPopulation.forEach((popCode) => {
-      let newEpisode;
-      let qdmDataElements;
-      if (cqlMap[popCode]) {
-        const definedPops = cqlMap[popCode];
-        const cqlPopulation = this.getPopulationFromIndex(definedPops, populationIndex, population, popCode);
+    for (const popCode in populationSet.populations) {
+      // Mongoose adds keys that start with '_' we do not care about these
+      if (popCode[0] !== '_') {
```
Comment (jbradl11) [inline lib/helpers/calculator_helpers.js]: See above comment

File: lib/helpers/calculator_helpers.js [hunk 10]
```diff
@@ -159,183 +151,162 @@ module.exports = class CalculatorHelpers {
...
+            if (qdmDataElement.id != null) {
+              // if an episode has already been created set the result for the population to 1
+              if (episodeResults[qdmDataElement.id.value]) {
+                episodeResults[qdmDataElement.id.value][popCode] = 1;
+                // else create a new episode using the list of all popcodes for the population
+              } else {
+                newEpisode = {};
+                for (const pc in populationSet.populations) {
+                  // Mongoose adds keys that start with '_' we do not care about these
+                  if (pc[0] !== '_') {
```
Comment (jbradl11) [inline lib/helpers/calculator_helpers.js]: See above comment.

File: lib/helpers/calculator_helpers.js [hunk 11]
```diff
@@ -159,183 +151,162 @@ module.exports = class CalculatorHelpers {
...
-              if (episodeResults[episodeId].values != null) {
-                episodeResults[episodeId].values.push(resultValue);
-              } else {
-                episodeResults[episodeId].values = [resultValue];
+            // else create a new episodeResult structure
+          } else {
+            const newEpisode = {};
+            for (const pc in populationSet.populations) {
+              // Mongoose adds keys that start with '_' we do not care about these
+              if (pc[0] !== '_') {
```
Comment (jbradl11) [inline lib/helpers/calculator_helpers.js]: See above comment

File: lib/models/calculator.js [hunk 12]
```diff
@@ -69,64 +69,57 @@ module.exports = class Calculator {
...
-        mainLibraryVersion = elmLibrary.library.identifier.version;
-        mainLibraryIndex = index;
-      }
-    }
+    const allElm = CalculatorHelpers.setValueSetVersionsToUndefined(_.clone(measure.cql_libraries.map(lib => lib.elm)));
+    // Find the main library (the library that is the "measure")
+    const mainLibraryElm = allElm.find(elm => elm.library.identifier.id === measure.main_cql_library);
 
-    const observations = measure.observations;
+    const observations = measure.population_sets[0].observations;
```
Comment (jbradl11) [inline lib/models/calculator.js]: Does this mean only the first population set will have `observations`?

Comment (mayerm94) [reply lib/models/calculator.js]: all population_sets have the same observations set on them

Comment (losborne) [reply lib/models/calculator.js]: Related to https://github.com/projecttacoma/cqm-execution/pull/23#discussion_r247580083 above

File: lib/helpers/measure_helpers.js [hunk 13]
```diff
@@ -355,22 +343,10 @@ module.exports = class MeasureHelpers {
...
-      if (curStatement.name === statementName) {
-        statement = curStatement;
-      }
-    }
-
-    if (library != null && statement != null) {
+  static isStatementFunction(library, statementName) {
+    // find the library and statement in the elm
+    const statement = library.elm.library.statements.def.find(def => def.name === statementName);
+    if (statement != null) {
```
Comment (jbradl11) [inline lib/helpers/measure_helpers.js]: When `find` does not return results, it returns `undefined`. Does that mean this should check for `statement != undefined`?

File: lib/helpers/results_helpers.js [hunk 14]
```diff
@@ -139,7 +127,12 @@ module.exports = class ResultsHelpers {
       statementRelevance[libraryName][statementName] === 'FALSE'
     ) {
       statementRelevance[libraryName][statementName] = relevant ? 'TRUE' : 'FALSE';
-      return Array.from(cqlStatementDependencies[libraryName][statementName]).map(dependentStatement =>
+      const library = cqlStatementDependencies.find(lib => lib.library_name === libraryName);
+      const statement = library.statement_dependencies.find(stat => stat.statement_name === statementName);
+      if (!statement.statement_references) {
```
Comment (jbradl11) [inline lib/helpers/results_helpers.js]: In the off chance `statement` is `undefined`, can we add a check to make sure it exists before we attempt to access `statement_references`?

File: lib/helpers/measure_helpers.js (hunk 13 above)
Comment (mayerm94) [reply lib/helpers/measure_helpers.js]: image

Comment (mayerm94) [reply lib/helpers/measure_helpers.js]: Should be fine

File: lib/models/calculator.js [hunk 15]
```diff
@@ -14,8 +15,8 @@ module.exports = class Calculator {
   /**
    * Generate calculation results for patients against a CQL measure.
    *
-   * @param {Measure} measure - The measure population to calculate on.
-   * @param {Array} patients - The array of QDM patients to run calcs on.
+   * @param {Object|Measure} measure - The measure population to calculate on.
+   * @param {Array<Object|QDMPatient>} patients - The array of QDM patients to run calcs on.
    * @param {Hash} valueSetsByOid - all ValueSets relevant to the measure.
```
Comment (hossenlopp) [inline lib/models/calculator.js]: `valueSetsByOid` is a misnomer.

**Comment by losborne** [comment]
Backend excel regression test using this branch came out clean.
